optuna/storages/_rdb/alembic/versions/v3.0.0.d.py

```diff
@@ -0,0 +1,170 @@
+"""Handle inf/-inf for trial_values table.
+
+Revision ID: v3.0.0.d
+Revises: v3.0.0.c
+Create Date: 2022-06-02 09:57:22.818798
+
+"""
+import enum
+
+import numpy as np
+from alembic import op
+import sqlalchemy as sa
+from sqlalchemy.exc import SQLAlchemyError
+from sqlalchemy.ext.declarative import declarative_base
+from sqlalchemy import orm
+from typing import Optional
+from typing import Tuple
+
+
+# revision identifiers, used by Alembic.
+revision = "v3.0.0.d"
+down_revision = "v3.0.0.c"
+branch_labels = None
+depends_on = None
+
+
+BaseModel = declarative_base()
+RDB_MAX_FLOAT = np.finfo(np.float32).max
+RDB_MIN_FLOAT = np.finfo(np.float32).min
+
+
+FLOAT_PRECISION = 53
+
+
+class TrialValueModel(BaseModel):
+    class TrialValueType(enum.Enum):
+        FINITE = 1
+        INF_POS = 2
+        INF_NEG = 3
+
+    __tablename__ = "trial_values"
+    trial_value_id = sa.Column(sa.Integer, primary_key=True)
+    value = sa.Column(sa.Float(precision=FLOAT_PRECISION), nullable=True)
+    value_type = sa.Column(sa.Enum(TrialValueType), nullable=False)
+
+    @classmethod
+    def value_to_stored_repr(
+        cls,
+        value: float,
+    ) -> Tuple[Optional[float], TrialValueType]:
+        if value == float("inf"):
+            return (None, cls.TrialValueType.INF_POS)
+        elif value == float("-inf"):
+            return (None, cls.TrialValueType.INF_NEG)
+        else:
+            return (value, cls.TrialValueType.FINITE)
+
+    @classmethod
+    def stored_repr_to_value(cls, value: Optional[float], float_type: TrialValueType) -> float:
+        if float_type == cls.TrialValueType.INF_POS:
+            assert value is None
+            return float("inf")
+        elif float_type == cls.TrialValueType.INF_NEG:
+            assert value is None
+            return float("-inf")
+        else:
+            assert float_type == cls.TrialValueType.FINITE
+            assert value is not None
+            return value
+
+
+def upgrade():
+    bind = op.get_bind()
+
+    sa.Enum(TrialValueModel.TrialValueType).create(bind, checkfirst=True)
+
+    # MySQL and PostgreSQL supports DEFAULT clause like 'ALTER TABLE <tbl_name>
+    # ADD COLUMN <col_name> ... DEFAULT "FINITE"', but seemingly Alembic
+    # does not support such a SQL statement. So first add a column with schema-level
+    # default value setting, then remove it by `batch_op.alter_column()`.
+    with op.batch_alter_table("trial_values") as batch_op:
+        batch_op.add_column(
+            sa.Column(
+                "value_type",
+                sa.Enum("FINITE", "INF_POS", "INF_NEG", name="trialvaluetype"),
+                nullable=False,
+                server_default="FINITE",
+            ),
+        )
+    with op.batch_alter_table("trial_values") as batch_op:
+        batch_op.alter_column("value_type", server_default=None)
```

Comment:
I thought MySQL requires `existing_type` and `existing_nullable`, but the current code worked as expected in my environment.
c.f., https://alembic.sqlalchemy.org/en/latest/ops.html#alembic.operations.Operations.alter_column.params.existing_type


I used the `mysql:8.0.25` docker image with [optuna-e2e](https://github.com/himkt/optuna-e2e).

```sql
CREATE TABLE public.trial_values (
    trial_value_id integer NOT NULL,
    trial_id integer NOT NULL,
    objective integer NOT NULL,
    value double precision,
    value_type public.trialvaluetype NOT NULL
);

```

Reply:
I also confirmed that the migration has passed, but I cannot understand the consistency with the documentation.
 🤔  

We have already missed `existing_nullable` in the migration script [here](https://github.com/optuna/optuna/blob/master/optuna/storages/_rdb/alembic/versions/v2.6.0.a_.py#L23).

Reply:
I also updated the migration code in v3.0.0.c, but left all previous code.

Reply:
Thank you for your update. Honestly, I prefer separated PRs to change the different migration scripts. But, it sounds reasonable for this case due to the release schedule. I checked both trial model and trial intermediate model and I'll share the result.